### PR TITLE
[TWILL-145] Potential race condition when restart all is called for a TwillRunnable

### PR DESCRIPTION
**Pull request by hsaputra**
If restart all instances is requested for a TwillRunnable then there could be race condition to check
provisioned and container requests that could exit the TwillApplication.

This PR containes changes:
-) Change the container requests to be ConcurrentLinkedQueue since it is accessed by multiple threads.
-) Add new volatile flag in RunnableContainerRequest to indicate whether it is ready to be provisioned.
-) Move up adding container requests for restart before removing.
-) Remove execution of restart to thread in the add instances executor.

### CODE REVIEW

**Comment by hsaputra** [comment]
Ran 8 Travis builds successfully.

File: twill-yarn/src/main/java/org/apache/twill/internal/appmaster/ApplicationMasterService.java [hunk 1]
```diff
@@ -835,35 +850,57 @@ private boolean handleRestartRunnablesInstances(final Message message, final Run
...
+        runnableContainerRequests.add(containerRequest);
+
+        for (int instanceId : instancesToRemove) {
+          LOG.debug("Remove instance {} for runnable {}", instanceId, runnableName);
+          try {
+            runningContainers.removeById(runnableName, instanceId);
+          } catch (Exception ex) {
+            // could be thrown if the container already stopped.
+            LOG.info("Exception thrown when stopping instance {} probably already stopped.", instanceId);
+          } finally {
```
Comment (chtyim) [inline twill-yarn/src/main/java/org/apache/twill/internal/appmaster/ApplicationMasterService.java]: Should this finally be outside of the for loop? There is only one instance of `constainerRequest` and it seems unnecessary to set the ready to true for every loop but rather do it just once when the looping finished.

**Comment by chtyim** [comment]
Just one comment, rest LGTM.

File: twill-yarn/src/main/java/org/apache/twill/internal/appmaster/ApplicationMasterService.java (hunk 1 above)
Comment (hsaputra) [inline twill-yarn/src/main/java/org/apache/twill/internal/appmaster/ApplicationMasterService.java]: Sounds, good make the change.

**Comment by hsaputra** [comment]
@chtyim I just updated the PR. Thanks!

**Comment by chtyim** [comment]
LGTM. Please squash.

**Comment by hsaputra** [comment]
Thanks, just squashed it

**Comment by chtyim** [comment]
Merged. I forget to update the commit message to close this PR. Please close it manually.

**Comment by hsaputra** [comment]
Closed per @chtyim

**Pull request closed without merging**